src/pexprs-madlibs.js

```diff
@@ -0,0 +1,209 @@
+}
+
+pexprs.Alt.prototype.generateExample = function(grammar, examples, inSyntacticContext, actuals) {
+  // items -> termExamples
+  var termExamples = this.terms.map(function(term) {
+    return term.generateExample(grammar, examples, inSyntacticContext, actuals);
+  });
+
+  var categorizedExamples = categorizeExamples(termExamples);
+
+  var examplesNeeded = categorizedExamples.examplesNeeded;
+  var successfulExamples = categorizedExamples.successfulExamples;
+  var needHelp = categorizedExamples.needHelp;
+
+  var returnObj = {};
+
+  // Alt can contain both an example and a request for examples
+  if (successfulExamples.length > 0) {
+    var i = Math.floor(Math.random() * successfulExamples.length);
+    returnObj.example = successfulExamples[i];
+  }
+  if (examplesNeeded.length > 0) {
+    returnObj.examplesNeeded = examplesNeeded;
+  }
+  returnObj.needHelp = needHelp;
+
+  return returnObj;
+};
+
+// safe thing to do might just be to omit whitespace period?
+// for lexical rules, should we request their components?
+
+pexprs.Seq.prototype.generateExample = function(grammar, examples, inSyntacticContext, actuals) {
+  var factorExamples = this.factors.map(function(factor) {
+    return factor.generateExample(grammar, examples, inSyntacticContext, actuals);
+  });
+  var categorizedExamples = categorizeExamples(factorExamples);
+
+  var examplesNeeded = categorizedExamples.examplesNeeded;
+  var successfulExamples = categorizedExamples.successfulExamples;
+  var needHelp = categorizedExamples.needHelp;
+
+  var returnObj = {};
+
+  // in a Seq, all pieces must succeed in order to have a successful example
+  if (examplesNeeded.length > 0 || needHelp) {
+    returnObj.examplesNeeded = examplesNeeded;
+    returnObj.needHelp = needHelp;
+  } else {
+    returnObj.example = successfulExamples.join(inSyntacticContext ? ' ' : '');
+  }
+
+  return returnObj;
+};
+
+pexprs.Apply.prototype.generateExample = function(grammar, examples, inSyntacticContext, actuals) {
+  var returnObj = {};
+
+  var ruleName = this.substituteParams(actuals).toString();
+
+  if (!examples.hasOwnProperty(ruleName)) {
+    returnObj.examplesNeeded = [ruleName];
+  } else {
+    var relevantExamples = examples[ruleName];
+    var i = Math.floor(Math.random() * relevantExamples.length);
+    returnObj.example = relevantExamples[i];
+  }
+
+  return returnObj;
+};
+
+// assumes that terminal's object is always a string
+pexprs.Terminal.prototype.generateExample = function(grammar, examples, inSyntacticContext) {
+  return {example: this.obj};
+};
+
+pexprs.Range.prototype.generateExample = function(grammar, examples, inSyntacticContext) {
+  var rangeSize = this.to.charCodeAt(0) - this.from.charCodeAt(0);
+  return {example: String.fromCharCode(
+    this.from.charCodeAt(0) + Math.floor(rangeSize * Math.random())
+  )};
+};
+
+// these all fail for now
+pexprs.Not.prototype.generateExample = function(grammar, examples, inSyntacticContext) {
+  return {example: ''};
```

Comment:
Not and Lookahead generate the empty string as a reasonable default to attempt to generate a valid example (basically, they hope that whatever comes after works according to the encoded constraints). It would be worth it to make this smarter for future work